### PR TITLE
Enable xla:gpu autocast for bfloat16 if not restricted

### PR DESCRIPTION
**Pull request by yeounoh**
`torch_xla.amp.autocast` was not using `XLA:GPU` but `torch.cuda` (GPU) backend. We should use `XLA` devices for XLA tensors, and for running `torch_xla.amp.autocast`.

Clarification 1:
> AMP executes on XLA:GPU, unless it's bfloat16 in which case, it has to fall back to cuda and requires USE_CUDA=1. We should address it.

This pytorch/pytorch#109302 to update the dispatcher comment is not needed. This also addresses #5497.

### CODE REVIEW

**Comment by JackCaoG** [comment]
there were some histories behind this. We used upstream autocase for cuda because we want to reuse the GPU AMP rules. Are you saying currently AMP does not execute on XLA:GPU when enabled?

**Comment by JackCaoG** [comment]
we also had test for it https://github.com/pytorch/xla/blob/master/test/test_autocast.py

**Comment by JackCaoG** [comment]
I am pretty sure, at least in 2.0 that our autocast actually worked with XLA:GPU and executed on XLA:GPU

**Comment by yeounoh** [comment]
I will investigate why our tests didn't catch this issue and modify or add proper tests, before merging.

**Comment by yeounoh** [comment]
> I am pretty sure, at least in 2.0 that our autocast actually worked with XLA:GPU and executed on XLA:GPU

Yea, it did. I think this issue was surfaced becasue of `bfloat16`, and we were hard-coding the dtype `float16` for GPU back then.

**Comment by yeounoh** [comment]
> there were some histories behind this. We used upstream autocase for cuda because we want to reuse the GPU AMP rules. Are you saying currently AMP does not execute on XLA:GPU when enabled?

Got it, AMP executes on XLA:GPU, unless it's `bfloat16` in which case, it has to fall back to `cuda` and requires `USE_CUDA=1`. We should address it.

**Comment by baoleai** [comment]
> Got it, AMP executes on XLA:GPU, unless it's `bfloat16` in which case, it has to fall back to `cuda` and requires `USE_CUDA=1`. We should address it.

How is this issue resolved?

**Comment by yeounoh** [comment]
> > Got it, AMP executes on XLA:GPU, unless it's `bfloat16` in which case, it has to fall back to `cuda` and requires `USE_CUDA=1`. We should address it.
> 
> How is this issue resolved?

Hi @baoleai it was resolved, but making changes to our approach. Will let you know when the PR is ready.

**Comment by yeounoh** [comment]
GPU test CI run timed out at the end.... increasing the timeout. I was able to pass all the GPU tests locally, on my GPU/TPU boxes. Will wait for the new CI run.

**Comment by ManfeiBai** [comment]
> GPU test CI run timed out at the end.... increasing the timeout. I was able to pass all the GPU tests locally, on my GPU/TPU boxes. Will wait for the new CI run.

Hi, does the GPU CI test timed out at a test? didn't found the time out error in the last two commits' log

Pin update PR met similar issue and fixed by increased the timeout, do we want to try this too? https://github.com/pytorch/xla/pull/5527/commits/55a08239b0cbdd5daebd15efca0f978a9369a560

File: torch_xla/runtime.py
Comment (JackCaoG) [inline torch_xla/runtime.py:91]: lol this is smart

**Review by JackCaoG — APPROVED** [review]

**Comment by yeounoh** [comment]
> > GPU test CI run timed out at the end.... increasing the timeout. I was able to pass all the GPU tests locally, on my GPU/TPU boxes. Will wait for the new CI run.
> 
> 
> 
> Hi, does the GPU CI test timed out at a test? didn't found the time out error in the last two commits' log
> 
> 
> 
> Pin update PR met similar issue and fixed by increased the timeout, do we want to try this too? https://github.com/pytorch/xla/pull/5527/commits/55a08239b0cbdd5daebd15efca0f978a9369a560

I increased timeout and was able to pass as well. I am going to address the gpu test time sometime soon.

File: torch_xla/amp/autocast_mode.py
Comment (baoleai) [inline torch_xla/amp/autocast_mode.py:36]: Why does not support bfloat16 now? Previously torch_xla was able to support bfloat16 with `torch.cuda.is_available = True`.  Do you mean we have to compile PyTorch with USE_CUDA=0 after this commit? And in most scenarios, our PyTorch needs to support both XLA and GPUs simultaneously, i.e. with USE_CUDA=1.

Comment (yeounoh) [reply torch_xla/amp/autocast_mode.py:36]: If torch is already compiled with `USE_CUDA=1` and `torch.cuda.is_available == True` then it would run with `cuda` backend and supports `bfloat16` as the cuda backend allows (note that for some platforms this wouldn't be the case). 

For the context, XLA:GPU, note that we guide users to compile `torch` with `USE_CUDA=0` to run ops with XLA, not torch cuda backend -- [gpu guide](https://github.com/pytorch/xla/blob/master/docs/gpu.md#develop-pytorchxla-on-a-gpu-instance-build-pytorchxla-from-source-with-gpu-support).

We've casted bfloat16 to float16/3 on XLA:GPU, but this should resolve it https://github.com/pytorch/xla/pull/5598

Comment (baoleai) [reply torch_xla/amp/autocast_mode.py:36]: > then it would run with cuda backend and supports bfloat16 as the cuda backend allows (note that for some platforms this wouldn't be the case).

But here the type is forced to torch.float16, even I set dtype=torch.bfloat16.

**Pull request merged**
